
# Задание 2




In [2]:
from google.colab import files
uploaded = files.upload()

Saving data.csv to data.csv


In [0]:
import pandas as pd

In [0]:
data = pd.read_csv('data.csv')

In [5]:
data.tail(5)

item_id       user_id  ... image_top_1 deal_probability
138710  5da3c1a0b09c  d3a29dbacfa1  ...      1285.0          0.12205
138711  192562b5fb5a  10c6a12cff32  ...      1055.0          0.73760
138712  7177e036a9fe  5069d63c1e47  ...      1243.0          0.20123
138713  8d3fe2371954  5e563726eb59  ...       559.0          0.00000
138714  875a1dc9575a  b2b28b445fa4  ...       999.0          0.00000

[5 rows x 18 columns]

In [0]:
#data['description'].head(50)

In [0]:
import numpy as np
from sklearn.metrics import mean_squared_error

In [0]:
from sklearn.model_selection import train_test_split

In [0]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.svm import LinearSVR

In [0]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

In [0]:
data['text_len'] = data['description'].apply(len)
data['text_len_tokens'] = data['description'].apply(lambda x: len(x.split()))
data['number_of_caps'] = data['description'].apply(lambda x: len([ch for ch in x if ch.isupper()]))
data['number_of_nonalphanum'] = data['description'].apply(lambda x: len([ch for ch in x if not ch.isalnum()]))
data['number_of_products'] = data['description'].apply(lambda x: len([ch for ch in x if ch == '+']))
data['used'] = data['description'].apply(lambda x: len([word for word in x.split() if word == 'б/у']))
data['excl'] = data['description'].apply(lambda x: len([ch for ch in x if ch == '!']))
data['condition'] = data['description'].apply(lambda x: len([word for word in x.split() if word == 'идеальном' or word == 'идеальное']))
data['eco'] = data['description'].apply(lambda x: len([word for word in x.split() if word == 'эко' or word == 'eco']))

data['number_of_products'] - количество плюсов. Чем больше плюсов, тем больше сопутсвующих товаров и аксессуаров, и тем, возможно, больше вероятность покупки.

data['used'] - б/у. Наверняка, на этот показатель покупатель тоже обращает внимание. 

data['excl'] - восклицательные знаки. Как я думаю, может уменьшать вероятность покупки. 

data['condition'] - говорит о состоянии продукта. Не факт, что влияет на вероятность (хотя должен), потому что "в идеальном состоянии" уже своего рода клише. 

data['eco'] - слова типа Эко-, эко или eco. Сейчас эта характеристика, вроде как, важна для покупателя (или, по крайней мере, важнее, чем раньше). 

In [0]:
X = data[['used', 'text_len', 'text_len_tokens', 'number_of_caps', 'number_of_nonalphanum', 'number_of_products', 'excl', 'condition', 'eco']].values
y = data.deal_probability.values

In [14]:
X.shape

(138715, 9)

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [16]:
regressor = Ridge(alpha=1)
regressor.fit(X_train, y_train)
preds = regressor.predict(X_test)
np.sqrt(mean_squared_error(y_test, preds))

0.26335659420046176

Новые характеристики улучшают качество, но совсем уж незначительно (примерно 0.00002 процента). Возможно, больше влияют какие-то метахарактеристики (типа размера текста, наличия или отсутсвия латиницы и т.д.). Видимо, интуитивно понятная логика в таких задачах малоприменима. 

In [0]:
from sklearn.linear_model import SGDRegressor

In [0]:
from scipy.sparse import hstack

In [0]:
cv = CountVectorizer(max_features=500, min_df=3, max_df=0.1, ngram_range=(1,2))
tfidf = TfidfVectorizer(max_features=500, min_df=3,analyzer='char', ngram_range=(3,3), max_df=0.1)

X_cv = cv.fit_transform(data.description)
X_tfidf = tfidf.fit_transform(data.description)

X = hstack([X_cv, X_tfidf])

In [113]:
X_cv.shape, X_tfidf.shape

((138715, 500), (138715, 500))

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [115]:
regressor = SGDRegressor(max_iter=100, alpha=1)
regressor.fit(X_train, y_train)
preds = regressor.predict(X_test)
np.sqrt(mean_squared_error(y_test, preds))

0.2625618984489414

Качество, пусть и не сильно, но повысилось. У векторизаторов уменьшил максимальное количество фич (так как значимых показателей в таких текстах, думаю, не очень много, и не хочется переобучить), min_df (потому что могут быть опечатки, и хотя бы три случая одинакового написания говорят о том, что слово может быть значимым) и max_df (потому что слова, которые ни на что не влияют (продам, куплю), встречаются часто).

У регрессора увеличил количетво итераций в два раза и подкрутил регуляризацию (что, кстати, более заметно повлияло на качество). 

Общая точность повысилась до 0.26256

